### ISSUE TITLE
`X-Forwarded-Host` doesn't support setting a port

### ISSUE THREAD

**Issue by ignasi35** [akka/akka-http#2191]
According to the [`Forwarded` spec](https://tools.ietf.org/html/rfc7239#section-5.3):

> The syntax for a "host" value, after potential quoted-string
   unescaping, MUST conform to the Host ABNF described in Section 5.4 of
   [RFC7230].

Then, Mozilla's MDN [documents the `X-Forwarded-Host` syntax](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Host) ambiguously. First is states:

> The X-Forwarded-Host (XFH) header is a de-facto standard header for identifying the original host requested by the client in the Host HTTP request header.

which seems to imply both `Host` and `X-Forwarded-Host` use the same syntax (as specified in [`Forwarded`](https://tools.ietf.org/html/rfc7239#section-5.3)). But then on the [`Syntax` section](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Host#Syntax) it's specified as:

> X-Forwarded-Host: <host>

which is different from the [MDN spec for `Host`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Host#Syntax):

> Host: <host>:<port>

The [current implementation in Akka-HTTP](https://github.com/akka/akka-http/blob/9cf194694e0459ceec75e2c88cfcdd45864b8aa2/akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/headers.scala#L1017-L1029) uses the `<host>` syntax and cites the MDN page linked above (which can be interpreted in different ways).

`X-Forwarded-Host` is a _de facto_ standard so we could be debating forever. Per the [`Forwarded` spec](https://tools.ietf.org/html/rfc7239#section-5.3), I think `X-Forwarded-Host` should allow setting a port.

**Comment by ignasi35** [comment]
An example use case:

 - In Lagom¡s dev mode, the Service Gateway (aka Reverse Proxy) listens on port `:9000` and redirects requests to downstream services listening on random ports. Currently the processed request receive `X-Forwarded-Host: localhost` instead of `X-Forwarded-Host: localhost:9000`.

**Comment by jrudolph** [comment]
I agree that could be useful. The header class is currently still marked `@ApiMayChange` so we may be able to just do the change.

(I don't quite understand the need for that header, why would a reverse proxy rewrite the `Host` header in the first place and not just forward it to the backend server unchanged?)

**Comment by TimMoore** [comment]
@jrudolph this is discussed by the HTTP spec in [RFC 7230](https://tools.ietf.org/html/rfc7230#section-5.4)

> When a proxy receives a request with an absolute-form of
> request-target, the proxy MUST ignore the received Host header field
> (if any) and instead replace it with the host information of the
> request-target.  A proxy that forwards such a request MUST generate a
> new Host field-value based on the received request-target rather than
> forward the received Host field-value.

But I think in this spec, "proxy" always refers to a forward client-side proxy, so it's a bit ambiguous about the case of a reverse proxy. It looks like implementations vary in practice, with some preserving the original `Host` and others replacing it.

I'm not sure where the `X-Forwarded-Host` convention originated, but it looks like some implementations use a separate `X-Forwarded-Port` header (as well as an `X-Forwarded-Scheme` or `X-Forwarded-Proto` to distinguish HTTP vs. HTTPS).

- https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/x-forwarded-headers.html#x-forwarded-port
- https://docs.microsoft.com/en-us/azure/application-gateway/application-gateway-faq
- https://github.com/kubernetes/ingress-nginx/commit/102c056b67fc64ee4098a01bdb0e294f7c6c7a85

Anyway, it seems like maybe for a non-standard header, it should be lax about what's allowed as the value.

**Comment by jlprat** [comment]
The way I understood those X-Forwarded-* headers is the following:
`X-Forwarded-Host`: contains only the host
`X-Forwarded-Port`: contains only the port

My interpretation is, `Forwarded` header is complicated to parse, this is why this set of non standard headers were introduced, to separate each component on a different header.

**Comment by ignasi35** [comment]
> The way I understood those X-Forwarded-* headers is the following:
> X-Forwarded-Host: contains only the host
> X-Forwarded-Port: contains only the port

The problem with _de-facto_ headers is that you are free to pick and choose. [MDN doesn't acknowledge the existence of a `X-Forwarded-Port` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Host) and [Wikipedia uses a `host:port` syntax in the `X-Forwarded-Host` example](https://en.wikipedia.org/wiki/List_of_HTTP_header_fields#Common_non-standard_request_fields). But then [Heroku uses `X-F-Port` and not `X-F-Host`](https://devcenter.heroku.com/articles/http-routing#heroku-headers). 🤷‍♂️ 

I think @TimMoore's suggestion is the best path forward:

> Anyway, it seems like maybe for a non-standard header, it should be lax about what's allowed as the value.

**Comment by jlprat** [comment]
Yes, I think we could be lax with what's in that header
